### ISSUE TITLE
Consider migrating Spotless Gradle plugin to use Task Configuration Avoidance

### ISSUE THREAD

**Issue by jbduncan** [diffplug/spotless#269]
I'm in the middle of migrating my personal project [jupiter-collection-testers](https://github.com/jbduncan/jupiter-collection-testers) to use a new feature introduced in Gradle 4.9 called [Task Configuration Avoidance](https://docs.gradle.org/4.9/userguide/task_configuration_avoidance.html), over at [this feature branch](https://github.com/jbduncan/jupiter-collection-testers/tree/migrate-task-configuration-avoidance). However, I am currently stuck on the Spotless Gradle plugin, because even if I don't run `gradlew spotless[Check|Apply]`, the plugin seems to directly depend on and thus eagerly configure (create) a number of tasks which `spotless[Check|Apply]` needs but other tasks like `help` don't need.

The reason that I believe Spotless is eagerly configuring these tasks, whichever ones they are, is because when I comment out [these](https://github.com/jbduncan/jupiter-collection-testers/blob/migrate-task-configuration-avoidance/build.gradle.kts#L7) [lines](https://github.com/jbduncan/jupiter-collection-testers/blob/migrate-task-configuration-avoidance/build.gradle.kts#L93-L131), the command `gradlew help -Dorg.gradle.internal.tasks.stats` produces the following relatively small output,

```text
1 actionable task: 1 executed
Task counts: Old API 1, New API 58, total 59
Task counts: created 3, avoided 56, %-lazy 95

Task types that were created with the old API
class com.github.benmanes.gradle.versions.updates.DependencyUpdatesTask 1

Task types that were registered with the new API but were created anyways
class org.gradle.configuration.Help 1
class org.gradle.api.tasks.bundling.Jar 1
```

as opposed to the following larger output,

```text
1 actionable task: 1 executed
Task counts: Old API 12, New API 74, total 86
Task counts: created 75, avoided 11, %-lazy 13

Task types that were created with the old API
class org.gradle.api.DefaultTask 8
class com.diffplug.gradle.spotless.SpotlessTask 3
class com.github.benmanes.gradle.versions.updates.DependencyUpdatesTask 1

Task types that were registered with the new API but were created anyways
class org.gradle.api.tasks.compile.JavaCompile 23
class org.gradle.api.DefaultTask 13
class org.gradle.api.tasks.Delete 9
class org.gradle.language.jvm.tasks.ProcessResources 3
class org.gradle.api.plugins.quality.Pmd 3
class org.gradle.api.tasks.testing.Test 1
class org.gradle.kotlin.dsl.accessors.tasks.PrintAccessors 1
class org.gradle.api.tasks.javadoc.Javadoc 1
class org.gradle.plugins.ide.eclipse.GenerateEclipseJdt 1
class org.gradle.plugins.ide.idea.GenerateIdeaModule 1
class org.gradle.plugins.ide.eclipse.GenerateEclipseClasspath 1
class org.gradle.plugins.ide.eclipse.GenerateEclipseProject 1
class org.gradle.kotlin.dsl.accessors.tasks.UpdateProjectSchema 1
class org.gradle.configuration.Help 1
class org.gradle.plugins.ide.idea.GenerateIdeaWorkspace 1
class org.gradle.api.tasks.bundling.Jar 1
class org.gradle.plugins.ide.idea.GenerateIdeaProject 1
```

**Comment by jbduncan** [comment]
@nedtwigg I'm in the middle of playing around with the Spotless source code to see if I can resolve this issue myself, but I'm unsure how to build the Spotless plugin and then expose it to jupiter-collection-testers to see if my changes work.

Do you have any advice as to how I could do this?

**Comment by nedtwigg** [comment]
> how to build the Spotless plugin and then expose it to jupiter-collection-testers

I would setup `mavenLocal()` as a plugin repository in `jupiter-collection-testers`, and then in Spotless run `gradlew publishToMavenLocal`.

I think it *might* also work to use composite builds and add `include '../spotless'` in the `settings.gradle` for `jupiter-collection-testers`.

**Comment by jbduncan** [comment]
Thanks @nedtwigg! I'll give `mavenLocal()` and `gradlew publishToMavenLocal` a shot (and composite builds if needed). :)

**Comment by jbduncan** [comment]
Hmm, _small_ problem: since Task Configuration Avoidance is a Gradle 4.9 feature, if I try to swap out eager `Task`s for their lazy equivalent - `TaskProvider`s - in the Spotless source code, it causes tests like [`FreshMarkExtensionTest`](https://github.com/diffplug/spotless/blob/master/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/FreshMarkExtensionTest.java) to fail because those tests call `new GradleRunner().withVersion("2.14.1"). ... .build()`.

I've considered swapping changing [this line in `GradleIntegrationTest`](https://github.com/diffplug/spotless/blob/master/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/GradleIntegrationTest.java#L63) so all the relevant tests run `withGradleVersion("4.9")` instead of version "2.14.1", but I don't know if you'd like that idea @nedtwigg, since you've expressed in https://github.com/diffplug/spotless/issues/161 that you still ran some stuff on Gradle 2.14.1 and couldn't migrate to future versions.

I've also considered using reflection, but I don't know if it's appropriate or even where to begin with it, as my knowledge of reflection is still beginner-level.

@nedtwigg Thoughts?

**Comment by jbduncan** [comment]
Would something like JOOR's [`Reflect.compile(String, String)`](https://static.javadoc.io/org.jooq/joor/0.9.9/org/joor/Reflect.html#compile-java.lang.String-java.lang.String-) help here?

**Comment by nedtwigg** [comment]
I would update gradle for your local build to 4.9, and not worry about compatibility.  Then I would benchmark.

For gradle 4.9, how much faster is a build with task configuration avoidance than without it?  My suspicion is that Spotless will gain little from it, because we already went to great lengths with the `FormatterStep.createLazy` machinations to keep task configuration lazy.

If we need to drop support for 2.14, 3, or even 4.8 so that we can get a significant performance improvement, then that might be worth it.  But if we're looking at a 10ms perf gain on a 5,000ms build, then I don't think it's worth taking on the maintenance burden of supporting an incubating API.

If the performance gains are substantial, we can then evaluate whether we do reflection, drop compat, etc.

**Comment by jbduncan** [comment]
Cool, thanks for sharing your thoughts @nedtwigg! I'll see how far I can go with making Spotless depend on Gradle 4.9.

I'll profile things at a later date, but (as I'll explain in my next comment), I'll profile Spotless against my project jupiter-collection-testers as well, as I believe that will be the best indicator of whether it's worth supporting the incubating API.

**Comment by jbduncan** [comment]
**tl;dr:** I agree that Spotless itself won’t gain much from task configuration avoidance, but I think it'll help most in situations where users (such as myself) use Spotless with _other_ plugins that “plug” into the `check` task.

To explain what I mean, my project jupiter-collection-testers has a custom `buildSrc` plugin for running a code refactoring tool called [Refaster](http://errorprone.info/docs/refaster) against the project's Java source code. This Refaster plugin has two unique tasks: `refasterApply` and `refasterCheck`.

Like Spotless, these two tasks respectively apply code refactorings and check if the code is already refactored. Also like Spotless, `refasterCheck` makes itself a subtask of `check`, so each time I run `gradlew check` on my project, it also runs `refasterCheck`.

However, my plugin actually runs Refaster _n * k_ times every time `refaster[Apply|Check]` is run, where _n_ is the number of [_Refaster templates_](http://errorprone.info/docs/refaster), and _k_ is the number of Java source sets in my Gradle project. This is because:

1. Refaster can only run with one Refaster template at a time.
2. Refaster builds upon [error-prone](http://errorprone.info/), so it's basically a custom Java compiler that refactors code a side-effect. This means that each time Refaster runs, it needs not only the Java sources to refactor but the classpath needed to compile those sources as well. For this reason, my plugin runs Refaster once per Gradle source set, since each source set provides not just Java sources but those sources' classpath too.

Under the hood, my plugin creates one `JavaCompile` task for each run of Refaster. This means when I run `gradlew refasterCheck`, it's not just running one task, it's really running _(n * k) + 1_ tasks.

And since my project imports Spotless, it eagerly configures (creates) all these tasks each and every time, even if I’m running an unrelated task like `gradlew dependencies`. This is because when Spotless is imported, [it directly configures (and thus creates) `check`](https://github.com/diffplug/spotless/blob/master/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java#L101-L108), meaning all _its_ subtasks get eagerly created in my project too, including my Refaster plugin’s `refasterCheck` task and its `JavaCompile` sub-tasks. This, I suspect, takes up a bit of time.

Long story short, if Spotless can avoid configuring `check` directly and do it lazily via task configuration avoidance, then it would in theory let my project avoiding creating `refasterCheck` if I run e.g. `gradlew dependencies`, saving time.

**Comment by nedtwigg** [comment]
My vote is to close it.  Search will dig it up if it becomes relevant again.  Thanks for the hard work implementing and benchmarking, good to know the results :)

**Comment by jbduncan** [comment]
You're very welcome, and cheers @nedtwigg. :)

**Comment by jbduncan** [comment]
I'm re-opening this issue for now, to better remind me to run the benchmarks again, which I believe are now misleading after a [helpful comment](https://github.com/diffplug/spotless/pull/277#issuecomment-417591945) from @thc202 implied that I benchmarked the wrong Gradle command!

**Comment by litpho** [comment]
We have noticed a rather large difference when configuration a custom format (the difference with just java isn't as remarkable).
![scan_without_task_avoidance](https://user-images.githubusercontent.com/5185114/45085379-dcf59b00-b100-11e8-866f-b874298adf2e.png)
![scan_with_task_avoidance](https://user-images.githubusercontent.com/5185114/45085378-dcf59b00-b100-11e8-9720-a2b20d196816.png)
Changes in Spotless plugin
[spotless.patch.txt](https://github.com/diffplug/spotless/files/2352192/spotless.patch.txt)
Spotless configuration in target project
[spotless-config.txt](https://github.com/diffplug/spotless/files/2352189/spotless-config.txt)

**Comment by jbduncan** [comment]
Thank you @litpho, that's good to know! I'm pretty sure now that I just messed up with my earlier benchmarking, so re-benchmarking this PR is at the top of my open-source todo list.

And apologies for the overall lack of activity. I started a new job recently, so I'm not sure yet when I'll have the time and energy to continue working on this PR. But stay tuned, nonetheless. :)

**Comment by nedtwigg** [comment]
Just an aside, #348 revealed an oopsie that was causing Spotless' configuration times to be way higher than necessary.  With that fixed, and our built-in lazy evaluation, my skepticism that this is a worthwhile investment for Spotless is renewed.

**Comment by jbduncan** [comment]
Ah okay, thanks for the update @nedtwigg.

Since my interest in this issue and the PR I opened for it (https://github.com/diffplug/spotless/pull/277) dwindled quite a while ago, I'd be perfectly fine if you decided to close this issue again. :)

**Comment by nedtwigg** [comment]
Looks like grolifant added a [utility library for this](http://ysb33rorg.gitlab.io/grolifant/0.11/docs/product-documentation.html#_lazy_create_tasks).

**Comment by nedtwigg** [comment]
This issue and its PR's have been a bit of a time sink, so I'm closing this issue.  I'm happy to reopen iff you have a build where configuration avoidance would improve performance, with [profiling data](https://github.com/diffplug/spotless/pull/425#issuecomment-518524069) to back it up.  Spotless already has built-in lazy configuration, so I'm skeptical that we gain anything with this change.

If we ever need to release a `4.0` version of Spotless, sure, we'll definitely adopt the gradle built-in API for this.  But unless we can show that the built-in API is faster (of which I am skeptical), then there's no reason to pay the cost.

**Comment by jbduncan** [comment]
@nedtwigg Agreed. Let's wait until if or when task configuration avoidance becomes the only option available. :)

**Comment by davidmoidl** [comment]
Hi @nedtwigg,

since this issue already exists, I'm not opening a new one, but I still think that Spotless plugin not playing well with Gradle's configuration avoidance is an issue.

I recently stumbled upon this when I tried to use configuration avoidance to simplify our build script and realized it's not working. After some debugging I came to the very same conclusion as OP did  - Spotless plugin actually causes some tasks to get configured, while these tasks would not get configured in its absence. This doesn't seem right to me.

You see, I don't really care about how Spotless plugin is coded and whether or not it uses the configuration avoidance internally _for it's own_ stuff, but I do think that it _should not_ mess with other tasks which aren't associated with spotless in any way. Not only because it's not necessary, but because some people (such as myself) might actually want to benefit from configuration avoidance in some way and Spotless plugin breaks this.

I didn't do any in-depth analysis, but just a quick check of the source code, yet I found [this](https://github.com/diffplug/spotless/blob/master/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java#L115) part of the code to be the cause of the problem. Obviously, iterating over these tasks makes Gradle configure them even though it normally wouldn't (using Gradle 5.6, btw.). Knowing that, I came up with a workaround: setting `enforceCheck` to `false` on the Spotless extension and manually setting `spotlessCheck` task as a dependency of Gradle's built-in `check` task does the trick.

Since I haven't yet had the need to study Gradle's Plugin API, I can't tell how difficult it would be to fix this. Nevertheless, I think it should be fixed and wanted to bring this up as IMHO, this is not just a performance issue, but an issue in general.

**Comment by nedtwigg** [comment]
For the benefit of issue subs, this has been released (partially, see #444 for details) in `3.25.0`.

**Comment by IgorGanapolsky** [comment]
How can we set up Spotless in Gradle KTS scripts?  It isn't allowing us to apply the `spotless` block.

**Comment by nedtwigg** [comment]
@IgorGanapolsky Not enough info to help - make a new issue with more context (see issue template)
